src/main/java/net/neoforged/neoforge/capabilities/BlockCapabilityCache.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities;
+
+import java.util.Objects;
+import java.util.function.BooleanSupplier;
+import net.minecraft.core.BlockPos;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.level.Level;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * A cache for block capabilities, to be used to track capabilities at a specific position, with a specific context.
+ * <p>
+ * The cache is invalidated when the level is notified of a change via {@link Level#invalidateCapabilities(BlockPos)}.
+ * <p>
+ * Instances are automatically cleared by the garbage collector when they are no longer in use.
+ */
+public final class BlockCapabilityCache<T, C> {
+    /**
+     * Creates a new cache instance and register it to the level.
+     *
+     * @param capability the capability
+     * @param level      the level
+     * @param pos        the position
+     * @param context    extra context for the query
+     */
+    public static <T, C> BlockCapabilityCache<T, C> create(BlockCapability<T, C> capability, ServerLevel level, BlockPos pos, C context) {
+        return create(capability, level, pos, context, () -> true, () -> {});
+    }
+
+    /**
+     * Creates a new cache instance with an invalidation listener, and register it to the level.
+     *
+     * @param capability           the capability
+     * @param level                the level
+     * @param pos                  the position
+     * @param context              extra context for the query
+     * @param isValid              a function to check if the listener still wants to receive notifications.
+     *                             A typical example would be {@code () -> !this.isRemoved() } for a block entity
+     *                             that should not receive invalidation notifications anymore once it is removed.
+     * @param invalidationListener the invalidation listener. Will be called whenever the capability of the cache might have changed.
+     */
+    public static <T, C> BlockCapabilityCache<T, C> create(BlockCapability<T, C> capability, ServerLevel level, BlockPos pos, C context, BooleanSupplier isValid, Runnable invalidationListener) {
+        Objects.requireNonNull(capability);
+        Objects.requireNonNull(isValid);
+        Objects.requireNonNull(invalidationListener);
+        pos = pos.immutable();
+
+        var cache = new BlockCapabilityCache<>(capability, level, pos, context, isValid, invalidationListener);
+        level.registerCapabilityListener(pos, cache.listener);
+        return cache;
+    }
+
+    private final BlockCapability<T, C> capability;
+    private final ServerLevel level;
+    private final BlockPos pos;
+    private final C context;
+
+    private boolean cacheValid = false;
+    @Nullable
+    private T cachedCap = null;
+
+    private boolean canQuery = true;
```

Comment:
I'm not sure I understand this... `canQuery` was mostly meant as a guard to make sure that the cache is not being queried while the invalidation listener runs. What benefit is there to controlling it?

Reply:
My wording wasn't the best sorry. I didn't mean exposing a way to modify `canQuery`, I just meant exposing a way to get the value of `canQuery`. And the use case I had in mind was if a mod is caching their `BlockCapabilityCache` (so that they can actually use it rather than basically just querying the level each time) and their `isValid` becomes false (and then potentially true again)... Currently there isn't a super good way to clear out their reference to the cache when retrieving it and getting a new one, unless `isValid` is sort of meant to have side effects when `false`. If `canQuery` was exposed as a getter, then mods could lazily replace their `BlockCapabilityCache` reference with a fresh "good" one.

Though also take the above with a minor grain of salt, as it could be I didn't fully understand a crash I was getting and thinking back on it, it could be related to attempting to query the cache (or more accurately get the value for it again to see if it is present or not) during the invalidation listener... Though I guess that sort of goes inline with your comment, which brings up the question... If `isValid` is true, what reason do we have for wanting to prevent querying the capability from the invalidation listener? As given it is inside the actual level listener, the capability should already have been replaced in level just before the invalidation of the previous value was fired... And my use case for checking if the capability is present during invalidation is for updating connections in a pipe network (the invalidation listener is used to check if a connection is still valid, or is now valid, which necessitates querying if the capability is actually present).

Reply:
The idea is that as soon as `isValid` returns false, the cache is removed from the level anyway. One should create a new cache in this case, I think. `isValid` should be seen as "definitely finished listening" rather than "currently not listening", IMO.

Reply:
> If isValid is true, what reason do we have for wanting to prevent querying the capability from the invalidation listener?

The idea is that the block firing the invalidation notification might not be in a state where it is ready to accept another query immediately. What I had in mind was that one would not react to invalidation notifications immediately, but rather "enqueue" a check for later. Happy to discuss your use case more, as this might be too restrictive.

Reply:
I am marking this as resolved. Most of this actually ended up being from confusion about a crash I was running into that was mainly related to me not being sure which of two methods of mine I should use. Using the delayed one resolves part of this. And the second part that was causing me issues I think was related to weirdness that was previously occurring due to block entity removal not invalidating capabilities, and thus I believe causing some weird things to happen on my end.